man/io_uring_internals.7

```diff
@@ -0,0 +1,225 @@
+.TH io_uring_internals 7 2024-10-5 "Linux" "Linux Programmer's Manual"
+.SH NAME
+io_uring_internals
+.SH SYNOPSIS
+.nf
+.B "#include <linux/io_uring.h>"
+.fi
+.PP
+.SH DESCRIPTION
+.PP
+.B io_uring
+is a linux specific, asynchronous API that allows the submission of requests to
+the kernel that are typically otherwise performed via a syscall. Requests are
+passed to the kernel via a shared ring buffer the
+.I Submission Queue
+(SQ) and completion notifications are passed back to the application via the
+.I Completion Queue
+(CQ). An important detail here is that after a request has been submitted to
+the kernel some CPU time has to be spent in kernel space to perform the
+required submission and completion related tasks.
+The mechanism used to provide this CPU time, as well as what process does so
+and when is different in
+.I io_uring
+than for the traditional API provided by regular syscalls.
+
+.PP
+.SH Traditional Syscall Driven I/O
+.PP
+For regular syscalls the CPU time for these tasks is directly provided by the
+process issuing the syscall, with the submission side tasks in kernel space
+being directly executed after the context switch. The time for completion
+related tasks is either also subsequently directly provided in the case of
+polled I/O. In the case of interrupt driven I/O the CPU time is provided,
+depending on the driver in question, by either the traditional top and bottom
+half IRQ approach or via threaded IRQ handling. The CPU time for completion
+tasks is thus in this case provided by the CPU on which the hardware
+interrupt arrives, as well as the CPU to which the dedicated kernel worker
+thread for the threaded IRQ handling gets scheduled, if that is used.
+
+.PP
+.SH The Submission Side Work
+.PP
+
+The tasks required in kernel space on the submission side are mostly checking
+the SQ for newly arrived SQEs, parsing and check them for validity and
+permissions and then passing them on to the responsible system, such as a
+block device driver. An important note here is that
+.I io_uring
+guarantees that the process of submitting the request to responsible subsystem
+and thus in this case the
+.IR io_uring_enter (2)
+syscall made to submit the new requests, will never block. However,
+.I io_uring
+relies on the capabilities of the responsible system to perform the submission
+without blocking.
+.I io_uring
+will first attempt to submit the request without blocking.
+If this fails, e.g. due to the respective system not supporting non-blocking
+submissions,
+.I io_uring
+will
+.I async punt
+the request, i.e. off-load these requests to the
+.I IO work queue
+(IO WQ) (see description below). 
+
+.PP
+.SH The Completion Side Work
+.PP
+
+The tasks required in kernel space on the completion side mostly come in the
```

Comment:
General theme - task is not the best word to use, because it implies a relationship to a thread/process. Not sure what's a better word to use here, just tossing it out there.

Reply:
Yeah true. Many of the alternatives are also overloaded. I guess "work" would be better. The heading already uses it.